executor/write.go

```diff
@@ -649,17 +646,15 @@ func (e *InsertExec) Next() (*Row, error) {
 	// If tidb_batch_insert is ON and not in a transaction, we could use BatchInsert mode.
 	batchInsert := e.ctx.GetSessionVars().BatchInsert && !e.ctx.GetSessionVars().InTxn()
 	rowCount := 0
-
 	for _, row := range rows {
 		if batchInsert && rowCount >= BatchInsertSize {
-			err = e.ctx.NewTxn()
-			if err != nil {
+			if err := e.ctx.NewTxn(); err != nil {
 				// We should return a special error for batch insert.
 				return nil, ErrBatchInsertFail.Gen("BatchInsert failed with error: %v", err)
 			}
-			txn = e.ctx.Txn()
 			rowCount = 0
 		}
+		txn := e.ctx.Txn()
```

Comment:
We only need to get a new txn after commit a batch of data. So it is better to move this back into the if statement.